File: mobile/android/app/src/main/java/com/findly/android/queue/worker/BackoffPolicy.kt
package com.findly.android.queue.worker

/**
 * Transient-flush-failure backoff (specs/009-device-runtime.md §9): exponential, 30 s initial,
 * doubling, **capped at the sync interval** ("never back off past the next natural capture").
 * WorkManager's own `setBackoffCriteria` implements this natively for the ≥15-minute path
 * (§3.1) — this pure calculator is for the §3.2 foreground-service path, whose self-rescheduling
 * timer isn't WorkManager-driven and so needs its own backoff bookkeeping.
 */
object BackoffPolicy {
    private const val INITIAL_DELAY_MILLIS = 30_000L
    private const val MAX_SHIFT = 20 // guards against Long overflow on a pathologically large attempt count

    /** [attempt] is 1-based (the first retry is attempt 1). */
    fun delayMillisForAttempt(attempt: Int, syncIntervalMinutes: Int): Long {
        require(attempt >= 1) { "attempt must be >= 1, was $attempt" }
        val capMillis = syncIntervalMinutes * 60_000L
        val shift = (attempt - 1).coerceAtMost(MAX_SHIFT)
        val exponential = INITIAL_DELAY_MILLIS * (1L shl shift)
        return exponential.coerceAtMost(capMillis)
    }
}

File: mobile/ios/FindlyKit/Sources/FindlyKit/LocationSensing/BackoffPolicy.swift
import Foundation

/// specs/009-device-runtime.md §9 — transient-flush-failure backoff: "exponential — 30 s initial,
/// doubling, capped at the sync interval (never back off past the next natural capture)." Pure,
/// standalone bookkeeping is needed on iOS because none of the three §3.4 triggers is a
/// WorkManager-style scheduler with native backoff support — `LocationSyncRunner` tracks its own
/// consecutive-failure attempt count and consults this before asking `BackgroundSyncScheduling` to
/// reschedule.
public enum BackoffPolicy {
    private static let initialDelaySeconds: TimeInterval = 30
    private static let maxShift = 20 // guards against overflow on a pathologically large attempt count

    /// `attempt` is 1-based (the first retry is attempt 1). Returns the delay in seconds.
    public static func delay(forAttempt attempt: Int, syncIntervalMinutes: Int) -> TimeInterval {
        precondition(attempt >= 1, "attempt must be >= 1, was \(attempt)")
        let capSeconds = TimeInterval(syncIntervalMinutes) * 60
        let shift = min(attempt - 1, maxShift)
        let exponential = initialDelaySeconds * pow(2, Double(shift))
        return min(exponential, capSeconds)
    }
}
